hardinfo/usb_util.c

```diff
@@ -0,0 +1,185 @@
+    if (s) {
+        memset(s, 0, sizeof(usbd));
+    }
+    return s;
+}
+
+void usbd_free(usbd *s) {
+    if (s) {
+        free(s->vendor);
+        free(s->product);
+        free(s->usb_version);
+        free(s->device_version);
+        free(s->dev_class_str);
+        free(s->dev_subclass_str);
+        free(s);
+    }
+}
+
+void usbd_list_free(usbd *s) {
+    usbd *n;
+    while(s != NULL) {
+        n = s->next;
+        usbd_free(s);
+        s = n;
+    }
+}
+
+/* returns number of items after append */
+int usbd_list_append(usbd *l, usbd *n) {
+    int c = 0;
+    while(l != NULL) {
+        c++;
+        if (l->next == NULL) {
+            if (n != NULL) {
+                l->next = n;
+                c++;
+            }
+            break;
+        }
+        l = l->next;
+    }
+    return c;
+}
+
+int usbd_list_count(usbd *s) {
+    return usbd_list_append(s, NULL);
+}
+
+char *_lsusb_lv(char *line, const char *prefix) {
+    if ( g_str_has_prefix(line, prefix) ) {
+        line += strlen(prefix) + 1;
+        return g_strstrip(line);
+    } else
+        return NULL;
+}
+
+int usb_get_device_lsusb(int bus, int dev, usbd *s) {
+    gboolean spawned;
+    gchar *out, *err, *p, *l, *t, *next_nl;
+    gchar *lsusb_cmd = g_strdup_printf("lsusb -s %d:%d -v", bus, dev);
+
+    s->bus = bus;
+    s->dev = dev;
+
+    spawned = g_spawn_command_line_sync(lsusb_cmd,
+            &out, &err, NULL, NULL);
+    g_free(lsusb_cmd);
+    if (spawned) {
+        if (strstr(err, "information will be missing")) {
+            s->user_scan = 1;
+        }
+        p = out;
+        while(next_nl = strchr(p, '\n')) {
+            strend(p, '\n');
+            g_strstrip(p);
+            if (l = _lsusb_lv(p, "idVendor")) {
+                s->vendor_id = strtol(l, NULL, 0);
+                if (t = strchr(l, ' '))
+                    s->vendor = strdup(t + 1);
+            } else if (l = _lsusb_lv(p, "idProduct")) {
+                s->product_id = strtol(l, NULL, 0);
+                if (t = strchr(l, ' '))
+                    s->product = strdup(t + 1);
+            } else if (l = _lsusb_lv(p, "MaxPower")) {
+                s->max_curr_ma = atoi(l);
+            } else if (l = _lsusb_lv(p, "bcdUSB")) {
+                s->usb_version = strdup(l);
+            } else if (l = _lsusb_lv(p, "bcdDevice")) {
+                s->device_version = strdup(l);
+            } else if (l = _lsusb_lv(p, "bDeviceClass")) {
+                s->dev_class = atoi(l);
+                if (t = strchr(l, ' '))
+                    s->dev_class_str = strdup(t + 1);
+            } else if (l = _lsusb_lv(p, "bDeviceSubClass")) {
+                s->dev_subclass = atoi(l);
+                if (t = strchr(l, ' '))
+                    s->dev_subclass_str = strdup(t + 1);
+            }
+            /* TODO: speed_mbs
+             * WISHLIST: interfaces, drivers */
+            p = next_nl + 1;
+        }
+        g_free(out);
+        g_free(err);
+        return 1;
```

Comment:
If a function is returning only 1 or 0, it should return a boolean value instead.